bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/InterleavedStorageRegenerateIndexOp.java

```diff
@@ -0,0 +1,199 @@
+
+    private final ServerConfiguration conf;
+    private final long ledgerId;
+
+    public InterleavedStorageRegenerateIndexOp(ServerConfiguration conf, long ledgerId) {
+        this.conf = conf;
+        this.ledgerId = ledgerId;
+    }
+
+    public void initiate(boolean dryRun) throws IOException {
+        LOG.info("Starting index rebuilding");
+
+        DiskChecker diskChecker = Bookie.createDiskChecker(conf);
+        LedgerDirsManager ledgerDirsManager = Bookie.createLedgerDirsManager(
+                conf, diskChecker, NullStatsLogger.INSTANCE);
+        LedgerDirsManager indexDirsManager = Bookie.createIndexDirsManager(
+                conf, diskChecker,  NullStatsLogger.INSTANCE, ledgerDirsManager);
+        EntryLogger entryLogger = new EntryLogger(conf, ledgerDirsManager);
+        final LedgerCache ledgerCache;
+        if (dryRun) {
+            ledgerCache = new DryRunLedgerCache();
+        } else {
+            ledgerCache = new LedgerCacheImpl(conf, new SnapshotMap<Long, Boolean>(),
+                                              indexDirsManager, NullStatsLogger.INSTANCE);
+        }
+
+        Set<Long> entryLogs = entryLogger.getEntryLogsSet();
+
+        int totalEntryLogs = entryLogs.size();
+        int completedEntryLogs = 0;
+        long startTime = System.nanoTime();
+
+        LOG.info("Scanning {} entry logs", totalEntryLogs);
+
+        AtomicLong firstEntry = new AtomicLong(Long.MAX_VALUE);
+        AtomicLong lastEntry = new AtomicLong(0);
+        AtomicLong numEntries = new AtomicLong(0);
+        for (long entryLogId : entryLogs) {
+            LOG.info("Scanning {}", entryLogId);
+            entryLogger.scanEntryLog(entryLogId, new EntryLogScanner() {
+                @Override
+                public void process(long ledgerId, long offset, ByteBuf entry) throws IOException {
+                    long entryId = entry.getLong(8);
+
+                    if (entryId < firstEntry.get()) {
+                        firstEntry.set(entryId);
+                    }
+                    if (entryId > lastEntry.get()) {
+                        lastEntry.set(entryId);
+                    }
+                    numEntries.incrementAndGet();
+
+                    // Actual location indexed is pointing past the entry size
+                    long location = (entryLogId << 32L) | (offset + 4);
+
+                    if (LOG.isDebugEnabled()) {
+                        LOG.debug("Rebuilding {}:{} at location {} / {}", ledgerId, entryId, location >> 32,
+                                location & (Integer.MAX_VALUE - 1));
+                    }
+
+                    if (!ledgerCache.ledgerExists(ledgerId)) {
+                        // set dummy master key and set fenced.
+                        // master key is only used on write operations
+                        ledgerCache.setMasterKey(ledgerId, new byte[0]);
```

Comment:
hmm. this would cause a problem, if we recover a ledger, but the writer comes in later and attempts to write entries, all the writes will be failing with auth failure. and the reader can't actually read them, right? do we have a test case to cover this negative case?

we might need to fix it or just add messages in the help messages of the tool, so people is aware of this situation.

Reply:
There's a little bit more hidden complexity here that I'm not sure how to resolve.

If a idx file has been deleted, but ledger metadata exists, when a client tries to open it and run recovery, it will see that it cannot read any entries, and close the ledger as empty.

I know DLog has a force option for this, but I think it would be better to provide another cli tool to allow users to mark the metadata from Closed -> In recovery, and then run the recovery again.

Reply:
It looks like we don't actually check the password on fencing, we just blindly accept what the client passes in :/

Reply:
> I think we need to pick up your effort to revamp the the tools, and 1) move them all into the same framework, 2) add tests & 3) add docs for each one.

let's pick up the tasks for #1000 

> If a idx file has been deleted, but ledger metadata exists, when a client tries to open it and run recovery, it will see that it cannot read any entries, and close the ledger as empty.

yeah. I think this is a rare corner case that we actuallly can do nothing for it.

> I know DLog has a force option for this, but I think it would be better to provide another cli tool to allow users to mark the metadata from Closed -> In recovery, and then run the recovery again.

sgtm

> It looks like we don't actually check the password on fencing, we just blindly accept what the client passes in :/

not sure we really need password for fencing, since it doesn't really append entries. and I think in general password should be removed eventually.

Reply:
Hmm, yes, I'll push my other updates for now, and leave out the password stuff, since it has no effect.